### PR TITLE
Mikrotik update cm

### PR DESCRIPTION
**Pull request by carlniger**
#1198 

@dBitech can you pull this down and run your tests again? Mostly this was cosmetic updates -- for consistency and to have less code, I collapsed the two classes into a Mikrotik base class since routeros/switchos were the same in your PR.  The only other big change was removing the static base_prompt you had set -- I did testing with a RouterOS VM and it seems to work just fine without it, but let us know if there is a reason for that to be there that I'm not seeing. Last thing was just renaming microtik -> Mikrotik.

@ktbyers let me know if I got the classes setup properly based on our chat

### CODE REVIEW

**Comment by carlniger** [comment]
Forgot to add testing output:

```
==================================================== test session starts =====================================================
platform darwin -- Python 3.7.2, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- /Library/Frameworks/Python.framework/Versions/3.7/bin/python3
cachedir: .pytest_cache
rootdir: /Users/carl/Dev/GitHub/netmiko
plugins: pylama-7.6.6, f5-sdk-3.0.21
collected 7 items

test_netmiko_config.py::test_ssh_connect PASSED                                                                        [ 14%]
test_netmiko_config.py::test_enable_mode PASSED                                                                        [ 28%]
test_netmiko_config.py::test_config_mode PASSED                                                                        [ 42%]
test_netmiko_config.py::test_exit_config_mode PASSED                                                                   [ 57%]
test_netmiko_config.py::test_command_set PASSED                                                                        [ 71%]
test_netmiko_config.py::test_commands_from_file PASSED                                                                 [ 85%]
test_netmiko_config.py::test_disconnect PASSED                                                                         [100%]

================================================== 7 passed in 7.70 seconds ==================================================
```

**Comment by carlniger** [comment]
And the show tests:

```
==================================================== test session starts =====================================================
platform darwin -- Python 3.7.2, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- /Library/Frameworks/Python.framework/Versions/3.7/bin/python3
cachedir: .pytest_cache
rootdir: /Users/carl/Dev/GitHub/netmiko
plugins: pylama-7.6.6, f5-sdk-3.0.21
collected 12 items

test_netmiko_show.py::test_disable_paging PASSED                                                                       [  8%]
test_netmiko_show.py::test_ssh_connect PASSED                                                                          [ 16%]
test_netmiko_show.py::test_ssh_connect_cm PASSED                                                                       [ 25%]
test_netmiko_show.py::test_send_command_timing PASSED                                                                  [ 33%]
test_netmiko_show.py::test_send_command_expect PASSED                                                                  [ 41%]
test_netmiko_show.py::test_base_prompt PASSED                                                                          [ 50%]
test_netmiko_show.py::test_strip_prompt PASSED                                                                         [ 58%]
test_netmiko_show.py::test_strip_command PASSED                                                                        [ 66%]
test_netmiko_show.py::test_normalize_linefeeds PASSED                                                                  [ 75%]
test_netmiko_show.py::test_clear_buffer PASSED                                                                         [ 83%]
test_netmiko_show.py::test_enable_mode PASSED                                                                          [ 91%]
test_netmiko_show.py::test_disconnect PASSED                                                                           [100%]

================================================= 12 passed in 21.61 seconds =================================================
```

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+
+class MikrotikBase(CiscoSSHConnection):
+    """Common Methods for Mikrotik RouterOS and SwitchOS"""
+
+    def __init__(self, **kwargs):
+        if kwargs.get("default_enter") is None:
+            kwargs["default_enter"] = "\r\n"
+
+        self.in_config_mode = True
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: Should default to False. Should be private variable i.e. `self._in_config_mode`.

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        return super(MikrotikBase, self).__init__(**kwargs)
+
+    def session_preparation(self, *args, **kwargs):
+        """Prepare the session after the connection has been established."""
+        self.ansi_escape_codes = True
+        # Clear the read buffer
+        self.write_channel(self.RETURN)
+        self.set_base_prompt()
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: Probably self.clear_buffer() at the end of this.

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _modify_connection_params(self):
+        """Append login options to username
+        c: disable console colors
+        e: enable dumb terminal mode
+        t: auto detect terminal capabilities
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: t: disable auto detect terminal capabilities (I assume this disables these)?

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        w511: set term width
+        h4098: set term height
+        """
+        self.username += "+cetw511h4098"
+
+    def _enter_shell(self):
+        """Already in shell."""
+        return ""
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: Shouldn't be needed--eliminate.

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _return_cli(self):
+        """The shell is the CLI."""
+        return ""
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: Shouldn't be needed--eliminate.

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        start the defaut return at len(self.get_prompt())+2*len(command)+1
+
+        :param command_string: The command string sent to the device
+        :type command_string: str
+
+        :param output: The returned output as a result of the command string sen
+        :type output: str
+        """
+        command_length = len(self.find_prompt()) + 2 * (len(command_string)) + 2
+        return output[command_length:]
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: Why is this needed?

File: netmiko/mikrotik/mikrotik_ssh.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        """No enable mode on RouterOS."""
+        return ""
+
+    def save_config(self, *args, **kwargs):
+        """No save command, all configuration is atomic"""
+        pass
+
+    def config_mode(self):
+        """No configuration mode on Microtik"""
+        self.in_config_mode = True
```
Comment (ktbyers) [inline netmiko/mikrotik/mikrotik_ssh.py]: Same with private attribute.

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 6 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: During development, it was found that the device returned the sent command twice, along with the router prompt. in the following format: Command\nRouterpromptCommand\n\n  This "solution" apeared to be the cleanest way to strip the commands.

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 1 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: Agreed, and in testing works as expected

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 2 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: Agreed, and in testing works as expected

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 5 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: Agreed, and in testing works as expected

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 7 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: Agreed, and in testing works as expected

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 4 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: Agreed, and in testing works as expected

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 3 above)
Comment (dBitech) [reply netmiko/mikrotik/mikrotik_ssh.py]: t: disables autodetection of terminal capabilities and sets height and witch as per below.

**Comment by carlniger** [comment]
updated per comments, re-ran tests:

```
platform darwin -- Python 3.7.2, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- /Library/Frameworks/Python.framework/Versions/3.7/bin/python3
cachedir: .pytest_cache
rootdir: /Users/carl/Dev/GitHub/netmiko
plugins: pylama-7.6.6, f5-sdk-3.0.21
collected 12 items

test_netmiko_show.py::test_disable_paging PASSED                                                                       [  8%]
test_netmiko_show.py::test_ssh_connect PASSED                                                                          [ 16%]
test_netmiko_show.py::test_ssh_connect_cm PASSED                                                                       [ 25%]
test_netmiko_show.py::test_send_command_timing PASSED                                                                  [ 33%]
test_netmiko_show.py::test_send_command_expect PASSED                                                                  [ 41%]
test_netmiko_show.py::test_base_prompt PASSED                                                                          [ 50%]
test_netmiko_show.py::test_strip_prompt PASSED                                                                         [ 58%]
test_netmiko_show.py::test_strip_command PASSED                                                                        [ 66%]
test_netmiko_show.py::test_normalize_linefeeds PASSED                                                                  [ 75%]
test_netmiko_show.py::test_clear_buffer PASSED                                                                         [ 83%]
test_netmiko_show.py::test_enable_mode PASSED                                                                          [ 91%]
test_netmiko_show.py::test_disconnect PASSED                                                                           [100%]

================================================= 12 passed in 21.59 seconds =================================================
```

```
==================================================== test session starts =====================================================
platform darwin -- Python 3.7.2, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- /Library/Frameworks/Python.framework/Versions/3.7/bin/python3
cachedir: .pytest_cache
rootdir: /Users/carl/Dev/GitHub/netmiko
plugins: pylama-7.6.6, f5-sdk-3.0.21
collected 7 items

test_netmiko_config.py::test_ssh_connect PASSED                                                                        [ 14%]
test_netmiko_config.py::test_enable_mode PASSED                                                                        [ 28%]
test_netmiko_config.py::test_config_mode PASSED                                                                        [ 42%]
test_netmiko_config.py::test_exit_config_mode PASSED                                                                   [ 57%]
test_netmiko_config.py::test_command_set PASSED                                                                        [ 71%]
test_netmiko_config.py::test_commands_from_file PASSED                                                                 [ 85%]
test_netmiko_config.py::test_disconnect PASSED                                                                         [100%]

================================================== 7 passed in 7.71 seconds ==================================================
```

Looking good!

File: netmiko/mikrotik/mikrotik_ssh.py (hunk 6 above)
Comment (carlniger) [reply netmiko/mikrotik/mikrotik_ssh.py]: @ktbyers per our slack convo I did a quick check and came to same solution @dBitech did - nice and simple.
